pos_accounting/src/pool/tests/simulation_tests.rs

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    chain::{DelegationId, Destination, OutPoint, OutPointSourceId, PoolId},
+    primitives::{Amount, Id, H256},
+};
+use crypto::random::{CryptoRng, Rng};
+use rstest::rstest;
+use test_utils::random::{make_seedable_rng, Seed};
+
+use super::create_pool_data;
+
+use crate::{
+    pool::{delegation::DelegationData, storage::PoSAccountingDB},
+    storage::in_memory::InMemoryPoSAccounting,
+    FlushablePoSAccountingView, PoSAccountingDelta, PoSAccountingOperations, PoSAccountingUndo,
+    PoSAccountingView,
+};
+
+fn random_outpoint0(rng: &mut impl Rng) -> OutPoint {
+    let source_id = OutPointSourceId::Transaction(Id::new(H256::random_using(rng)));
+    OutPoint::new(source_id, 0)
+}
+
+fn get_random_pool_id(rng: &mut impl Rng, storage: &InMemoryPoSAccounting) -> Option<PoolId> {
+    let all_pool_data = storage.all_pool_data();
+    if !all_pool_data.is_empty() {
+        let (pool_id, _) = all_pool_data.iter().nth(rng.gen_range(0..all_pool_data.len())).unwrap();
+        Some(*pool_id)
+    } else {
+        None
+    }
```

Comment:
Not sure it's much of an improvement but it seems you can use [bool::then](https://github.com/Cameri/nostream) here and a few other functions in this file:
```rust
(!all_pool_data.is_empty()).then(|| all_pool_data.iter().blah...())
```